bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/client/HomeConnectApiClient.java

```diff
@@ -0,0 +1,1134 @@
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.collections4.QueueUtils;
+import org.apache.commons.collections4.queue.CircularFifoQueue;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.homeconnect.internal.client.exception.ApplianceOfflineException;
+import org.openhab.binding.homeconnect.internal.client.exception.AuthorizationException;
+import org.openhab.binding.homeconnect.internal.client.exception.CommunicationException;
+import org.openhab.binding.homeconnect.internal.client.model.ApiRequest;
+import org.openhab.binding.homeconnect.internal.client.model.AvailableProgram;
+import org.openhab.binding.homeconnect.internal.client.model.AvailableProgramOption;
+import org.openhab.binding.homeconnect.internal.client.model.Data;
+import org.openhab.binding.homeconnect.internal.client.model.HomeAppliance;
+import org.openhab.binding.homeconnect.internal.client.model.HomeConnectRequest;
+import org.openhab.binding.homeconnect.internal.client.model.HomeConnectResponse;
+import org.openhab.binding.homeconnect.internal.client.model.Option;
+import org.openhab.binding.homeconnect.internal.client.model.Program;
+import org.openhab.core.auth.client.oauth2.OAuthClientService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+
+/**
+ * Client for Home Connect API.
+ *
+ * @author Jonas Brüstel - Initial contribution
+ *
+ */
+@NonNullByDefault
+public class HomeConnectApiClient {
+    private static final String ACCEPT = "Accept";
+    private static final String CONTENT_TYPE = "Content-Type";
+    private static final String BSH_JSON_V1 = "application/vnd.bsh.sdk.v1+json";
+    private static final MediaType BSH_JSON_V1_MEDIA_TYPE = requireNonNull(MediaType.parse(BSH_JSON_V1));
+    private static final int REQUEST_READ_TIMEOUT = 30;
+    private static final int VALUE_TYPE_STRING = 0;
+    private static final int VALUE_TYPE_INT = 1;
+    private static final int VALUE_TYPE_BOOLEAN = 2;
+    private static final int COMMUNICATION_QUEUE_SIZE = 50;
+
+    private final Logger logger;
+    private final OkHttpClient client;
+    private final String apiUrl;
+    private final Map<String, List<AvailableProgramOption>> availableProgramOptionsCache;
+    private final OAuthClientService oAuthClientService;
+    private final Queue<ApiRequest> communicationQueue;
+    private final JsonParser jsonParser;
+
+    public HomeConnectApiClient(OAuthClientService oAuthClientService, boolean simulated,
```

Comment:
Now that OpenHab depends on Java 11, the new Java 11 HTTP client could be used, slashing the size of the binding considerable.